lib/Onyx.js

```diff
@@ -449,25 +552,45 @@ function keyChanged(key, data, canUpdateSubscriber) {
  *     - triggers the callback function
  *
  * @private
- * @param {object} config
- * @param {object} [config.withOnyxInstance]
- * @param {string} [config.statePropertyName]
- * @param {function} [config.callback]
+ * @param {Object} mapping
+ * @param {Object} [mapping.withOnyxInstance]
+ * @param {String} [mapping.statePropertyName]
+ * @param {Function} [mapping.callback]
+ * @param {String} [mapping.selector]
  * @param {*|null} val
  * @param {String} matchedKey
  */
-function sendDataToConnection(config, val, matchedKey) {
+function sendDataToConnection(mapping, val, matchedKey) {
     // If the mapping no longer exists then we should not send any data.
     // This means our subscriber disconnected or withOnyx wrapped component unmounted.
-    if (!callbackToStateMapping[config.connectionID]) {
+    if (!callbackToStateMapping[mapping.connectionID]) {
+        return;
+    }
+
+    if (mapping.withOnyxInstance) {
+        // If the mapping has a selector, then the component's state must only be updated with the data
+        // returned by the selector.
+        if (mapping.selector) {
+            if (isCollectionKey(mapping.key)) {
+                const newData = reduceCollectionWithSelector(val, mapping.selector);
+                PerformanceUtils.logSetStateCall(mapping, null, newData, 'sendDataToConnection');
+                mapping.withOnyxInstance.setWithOnyxState(mapping.statePropertyName, newData);
+                return;
+            }
+
+            const newValue = getSubsetOfData(val, mapping.selector);
+            PerformanceUtils.logSetStateCall(mapping, null, newValue, 'sendDataToConnection');
+            mapping.withOnyxInstance.setWithOnyxState(mapping.statePropertyName, newValue);
+            return;
+        }
+
+        PerformanceUtils.logSetStateCall(mapping, null, val, 'sendDataToConnection');
+        mapping.withOnyxInstance.setWithOnyxState(mapping.statePropertyName, val);
```

Comment:
NAB this could probably be DRY'd up if we use a `let` variable for `newData`

```js
let newData;
if (mapping.selector) {
    if (collectionKey) {
        newData = ...;
    } else {
        newData = ...;
    }
} else {
    newData = ...;
}

PerformanceUtils.logSetStateCall(mapping, null, newData, 'sendDataToConnection');
mapping.withOnyxInstance.setWithOnyxState(mapping.statePropertyName, newData);
```

Reply:
Good idea!